code/app/src/main/java/com/example/bread/view/SignupPage.java

```diff
@@ -42,98 +57,197 @@ protected void onCreate(Bundle savedInstanceState) {
+            isValid = false;
+        }
+
+        // Additional validation
+        if (!TextUtils.isEmpty(username) && username.length() < MIN_USERNAME_LENGTH) {
+            usernameEditText.setError("Username must be at least " + MIN_USERNAME_LENGTH + " characters");
+            isValid = false;
+        }
+
+        if (!TextUtils.isEmpty(username) && !Pattern.matches(USERNAME_PATTERN, username)) {
+            usernameEditText.setError("Username can only contain letters, numbers, and underscores");
+            isValid = false;
+        }
+
+        if (!TextUtils.isEmpty(email) && !Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
+            emailEditText.setError("Enter a valid email address");
+            isValid = false;
+        }
+
+        if (!TextUtils.isEmpty(password) && password.length() < MIN_PASSWORD_LENGTH) {
+            passwordEditText.setError("Password must be at least " + MIN_PASSWORD_LENGTH + " characters");
+            isValid = false;
+        }
+
+        return isValid;
+    }
+
+    private void checkUsernameAvailability(String username, String email, String password, String firstName, String lastName) {
+        participantRepository.checkIfUsernameExists(username, exists -> {
+            if (exists) {
+                showLoading(false);
+                usernameEditText.setError("Username already exists");
+                usernameEditText.requestFocus();
+            } else {
+                createUserAccount(email, password, username, firstName, lastName);
+            }
+        }, e -> {
+            showLoading(false);
+            Log.e(TAG, "Failed to check if username exists", e);
+            Toast.makeText(SignupPage.this, "Network error. Please try again.", Toast.LENGTH_SHORT).show();
         });
     }
 
-    private void signUpUser(@NonNull String email, @NonNull String password, @NonNull OnSuccessListener<AuthResult> onSuccessListener, OnFailureListener onFailureListener) {
-        mAuth.createUserWithEmailAndPassword(email, password).addOnSuccessListener(onSuccessListener)
-                .addOnFailureListener(onFailureListener != null ? onFailureListener : e -> {
-                    Log.e("SignupPage", "Failed to sign up user", e);
-                    Toast.makeText(SignupPage.this, "Sign Up failed. Try again!", Toast.LENGTH_SHORT).show();
+    private void createUserAccount(String email, String password, String username, String firstName, String lastName) {
+        mAuth.createUserWithEmailAndPassword(email, password)
+                .addOnSuccessListener(this, authResult -> {
+                    FirebaseUser user = authResult.getUser();
+                    if (user != null) {
+                        updateUserProfile(user, username, email, firstName, lastName);
+                    } else {
+                        showLoading(false);
+                        Toast.makeText(SignupPage.this, "Account creation failed. Please try again.", Toast.LENGTH_SHORT).show();
```

Comment:
Try to keep message small as possible in the toast. Just a UI preference looks more clean that way. Applies to other toast messages as well.
```suggestion
                        Toast.makeText(SignupPage.this, "Account creation failed.", Toast.LENGTH_SHORT).show();
```